### PR TITLE
Tracker: more customization

### PR DESCRIPTION
**Pull request by tyanas**
significant change is here https://github.com/basisjs/basisjs/commit/c32bfd422f99f8710a63b00a60e1f330a25731b5#diff-a86cef1540e36750e43b0be3cdc18215 (2 lines)

everything else is preparations

### CODE REVIEW

File: src/basis/data.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 
  /**
-  * This namespace contains base classes and functions for data maintain.
+  * This namespace contains base classes and functions for data manipilation and data maintenance.
```
Comment (lahmatiy) [inline src/basis/data.js]: Changes are not related to PR

File: src/basis/dom.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 
  /**
-  * This namespace provides functions for DOM manupulations - transerval,
-  * node creation, moving and test nodes. Most of functions are compatible with
-  * native and simulated (object that generaly has properties like firsChild,
-  * lastChild, parentNode etc) DOM structures.
+  * This namespace provides functions for DOM manipulations - traversal,
```
Comment (lahmatiy) [inline src/basis/dom.js]: Changes are not related to PR

File: src/basis/tracker.js [hunk 3]
```diff
@@ -2,27 +2,78 @@
...
+/** @cut */ var namespace = 'basis.tracker';
+
 var hasOwnProperty = Object.prototype.hasOwnProperty;
 var eventUtils = require('basis.dom.event');
 var getComputedStyle = require('basis.dom.computedStyle').get;
 var getBoundingRect = require('basis.layout').getBoundingRect;
+
+/**
+ * The Map object is a simple key/value map. Any value (both objects and primitive values) may be used as either a key or a value.
+ * @external Map
```
Comment (lahmatiy) [inline src/basis/tracker.js]: `Map` is not used in this module. There is should be another way to define maps

File: src/basis/tracker.js [hunk 4]
```diff
@@ -2,27 +2,78 @@
...
+ * Event names can be like these: `click`, `focus`, `input`, `blur`, `success`, etc.
+ * eventMap.show will be a list of tracking objects, and taht list has `visibility` property,
+ * eventMap.show.visibility is an object: custom selector string --> boolean
+ * @private
+ * @type Map
+ */
 var eventMap = {};
 
 var VISIBLE_CHECK_INTERVAL = 250;
+var SHOW_EVENT = 'show';
```
Comment (lahmatiy) [inline src/basis/tracker.js]: Minor: redundant line between consts

File: src/basis/tracker.js [hunk 5]
```diff
@@ -117,18 +213,235 @@ function checkShow(){
...
+          return res;
+        }, []);
+
+        // if there is at least one `role-marker` attribute in the DOM [x]path for the clicked (or hovered or `event`ed) element
+        // then lets search a matching selector in our loaded track map
+        if (path.length)
+          selectorList.forEach(function(item){
+            var result = handleEventFor(path, item, event);
+            if (result)
+              if (result.async) {
```
Comment (lahmatiy) [inline src/basis/tracker.js]: `{` on next line

File: src/basis/tracker.js [hunk 6]
```diff
@@ -117,18 +213,235 @@ function checkShow(){
...
+                clearTimeout(inputTimeout);
+                inputTimeout = setTimeout(function(){
+                  track(result.dataToTrack);
+                }, INPUT_DEBOUNCE_TIMEOUT);
+              }
+              else {
```
Comment (lahmatiy) [inline src/basis/tracker.js]: `{` on next line

File: src/basis/tracker.js [hunk 7]
```diff
@@ -444,65 +686,16 @@ function addDispatcher(dispatcher, events, transformer){
...
-  return result.length ? result : false;
-}
-
-
 module.exports = {
-  parsePath: parsePath,
-  parseRole: parseRole,
-  stringifyPath: stringifyPath,
-  stringifyRole: stringifyRole,
+  /** @cut */ parseRole: parseRole,
```
Comment (lahmatiy) [inline src/basis/tracker.js]: What is the reason to make those functions private?

File: src/basis/tracker.js [hunk 8]
```diff
@@ -117,18 +213,235 @@ function checkShow(){
...
+
+  var data = basis.object.slice(item.data);
+  var INPUT_EVENTS = ['keyup', 'keydown', 'keypress', 'input']; // when a user is typing
+
+  if (INPUT_EVENTS.indexOf(event.type) != -1)
+  {
+    data.inputValue = event.target.value;
+
+    return {
+      async: true,
```
Comment (lahmatiy) [inline src/basis/tracker.js]: Minor: I think `async` -> `debounce` would be better

File: src/basis/tracker.js [hunk 9]
```diff
@@ -117,18 +213,235 @@ function checkShow(){
...
+ * Search for the first key in `obj` (`obj` can be a nested object)
+ * and replaces a value of the finded key with `value`
+ * private except for tests
+ * @param {Array.<RoleObject>} path
+ * @param {Object} item custom tracking data object from some tracking map
+ * @param {Object} item.data
+ * @param {Array.<string>} item.selector
+ * @param {string} item.selectorStr
+ * @param {DOMEvent} event
+ * @return {{ async: boolean, datatoTrack: TrackingDataObject }}
```
Comment (lahmatiy) [inline src/basis/tracker.js]: typo `datatoTrack` -> `dataToTrack`

File: src/basis/tracker.js [hunk 10]
```diff
@@ -117,18 +213,235 @@ function checkShow(){
...
+    data.inputValue = event.target.value;
+
+    return {
+      async: true,
+      dataToTrack: {
+        type: 'ui',
+        path: stringifyPath(path),
+        selector: stringifyPath(item.selector),
+        event: event.type,
+        data: typeof data.dataToTrackFn === 'function' ? data.dataToTrackFn(data, event) : data
```
Comment (lahmatiy) [inline src/basis/tracker.js]: 1. It would be better to normalize field on `item` create phase:
```js
item.data.dataToTrackFn = typeof ... === 'function' ? ... : basis.fn.$self;
```
In this case we don't need duplicate logic to check the field, just `data.dataToTrackFn(data, event)`.

2. I believe we need better name for `dataToTrackFn`

Comment (tyanas) [inline src/basis/tracker.js]: 1. when registering a selector (in `registrateSelector` function) there is no info about exact browser event. That is why `data` is transformed while handling an event.
Here we execute a custom transform function with the event. It adds desired dynamics to event tracking

2. Yep. There must be a way to put previous piece of text (1.) in a word :)

File: src/basis/tracker.js (hunk 7 above)
Comment (tyanas) [inline src/basis/tracker.js]: It's more like I didn't find a reason to leave it public. 
Moreover there is no tests for them for now.

File: src/basis/tracker.js (hunk 10 above)
Comment (lahmatiy) [inline src/basis/tracker.js]: 1. Function invocation should left as is here, but function normalization should be outside of `handleEventFor`:
```js
data: data.dataToTrackFn(data, event)
```

File: src/basis/tracker.js (hunk 7 above)
Comment (lahmatiy) [inline src/basis/tracker.js]: Tests absence is not the reason to make API private. It should be provate when there are no useful use cases in prod. Also if it's doesn't use anywhere there is no reason to add something to export.

Comment (tyanas) [inline src/basis/tracker.js]: `parseRole`, `stringifyPath` ,`stringifyRole` are used in devpanel only. 
And `isPathMatchSelector`, `setDeep` and `handleEventFor` are used in tests only.
That's why they are in  `module.exports`.
Such information is specified in jsdoc for those functions

And I do not know any use cases for them in production.

Comment (lahmatiy) [inline src/basis/tracker.js]: Let's left those methods public until we know for sure they are useless in prod.
